content/blog/2023-12-01-task-lifecycle.md

```diff
@@ -0,0 +1,572 @@
+call this by now too. These are considered 'idle tasks', which serve to run when
+no other tasks are available to run. They will spin on a check for pending work,
+entering an idle state each loop until they see pending tasks to run. They then
+call `__schedule()` in a loop (also conditional on pending tasks existing), and
+then return back to the idle loop once everything in the moment is handled.
+
+The `__schedule()` function is the main guts of the scheduler, which until now
+has seemed like some nebulous controller that's governing when and where our
+tasks run. In reality it isn't one isolated part of the system, but a function
+that a task calls when it decides to yield to any other waiting tasks. It starts
+by deciding which pending task should run (a whole can of worms right there),
+and then executing `context_switch()` if it changes from the current task.
+`context_switch()` is the point where the `current` task starts to change.
+Specifically, you can trace the changing of `current` (i.e., the PACA being
+updated with a new task pointer) to the following path
+
+```c
+context_switch()
+  switch_to()
+    __switch_to()
+      _switch()
+        do_switch_64
+          std	r6,PACACURRENT(r13)
+```
+
+One interesting consequence of tasks calling `context_switch()` is that the
+previous task is 'suspended'[^2] right where it saves its registers and puts in
+the new task's values. When it is woken up again at some point in the future it
+resumes right where it left off. So when you are reading the `__switch_to()`
+implementation, you are actually looking at two different tasks in the same
+function.
+
+[^2]: Don't forget that the entire concept of tasks is made up by the kernel:
+    from the hardware's point of view we haven't done anything interesting, just
+    changed some registers.
+
+But it gets even weirder: while tasks that put themselves to sleep here wake up
+inside of `_switch()`, new tasks being woken up for the first time start at a
+completely different location! So not only is the task changing, the `_switch()`
+call might not even return back to `__switch_to()`!
+
+
+## Conclusion
+
+And there you have it, everything<sup>_[citation needed]_</sup> you could ever
+need to know when getting started with tasks. Will you need to know this
+specifically? Hard to say. But hopefully it at least provides some useful
+pointers for understanding the execution model of the kernel.
+
+
+## Questions
+
+The following are some questions you might have (read: I had).
+
+### Do userspace tasks run in the kernel?
+
+Yes, 'userspace' tasks run in the kernel. We don't have dedicated tasks to swap
+out to. Thanks to address space quadrants we don't even need to change the
+active memory mapping: upon entry to the kernel we automatically start using the
+PID 0 mapping.
+
+
+### Where does a task get allocated a PID?
+
+Referring to the hardware PID, used for virtual memory translations, this is
```

Comment:
what about a software PID, which is what most people will think of when they read the question

Reply:
Added more context